File: Data Structures & Algorithms/max-area-of-island/submission-1.java
class Solution {
    private int[][] directions = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    public int maxAreaOfIsland(int[][] grid) {
        int ROWS = grid.length;
        int COLS = grid[0].length;

        int max = 0;

        for (int r=0; r<ROWS; r++) {
            for (int c=0; c<COLS; c++) {
                if (grid[r][c]==1) {
                    int current = currArea(grid, r, c);
                    if (max<current) {
                        max=current;
                    }
                }
            }
        }
        return max;
    }

    public int currArea(int[][] grid, int r, int c) {
        if (r<0 || c<0 || r>=grid.length || c>=grid[0].length || grid[r][c]==0) {
            return 0;
        }

        grid[r][c]=0;

        int area=1;

        for (int[] dir : directions) {
            area+=currArea(grid, r+dir[0], c+dir[1]);
        }

        return area;
    }
}

File: Data Structures & Algorithms/unique-paths-ii/submission-1.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        
        ROWS, COLS = len(obstacleGrid), len(obstacleGrid[0])

        cache = [[-1]*COLS for _ in range(ROWS)]

        def top_down(grid: list[list[int]], r, c, ROWS, COLS):
            if r==ROWS or c==COLS or grid[r][c]==1:
                return 0
            
            if r==ROWS-1 and c==COLS-1:
                return 1
            
            if cache[r][c]!=-1:
                return cache[r][c]

            cache[r][c]=(top_down(grid, r+1, c, ROWS, COLS)+top_down(grid,r,c+1,ROWS, COLS))

            return cache[r][c]

        
        return top_down(obstacleGrid, 0, 0, ROWS, COLS)